src/App.tsx

```diff
@@ -0,0 +1,800 @@
+    setColorRampMin(canv.getColorMapRangeMin());
+    setColorRampMax(canv.getColorMapRangeMax());
+  }, [canv]);
+
+  const getFeatureValue = useCallback(
+    (id: number): number => {
+      if (!featureName || !dataset) {
+        return -1;
+      }
+      // Look up feature value from id
+      return dataset.getFeatureData(featureName)?.data[id] || -1;
+    },
+    [featureName, dataset]
+  );
+
+  const onMouseMove = useCallback(
+    (event: MouseEvent): void => {
+      if (!dataset) {
+        return;
+      }
+      const id = canv.getIdAtPixel(event.offsetX, event.offsetY);
+      if (id === BACKGROUND_ID) {
+        // Ignore background pixels
+        return;
+      }
+      setHoveredId(id);
+    },
+    [dataset, canv]
+  );
+
+  const onMouseLeave = useCallback((_event: MouseEvent): void => {
+    setHoveredId(null);
+  }, []);
+
+  useEffect(() => {
+    canv.domElement.addEventListener("mousemove", onMouseMove);
+    canv.domElement.addEventListener("mouseleave", onMouseLeave);
+    return () => {
+      canv.domElement.removeEventListener("mousemove", onMouseMove);
+      canv.domElement.removeEventListener("mouseleave", onMouseLeave);
+    };
+  }, [onMouseMove, onMouseLeave, canv]);
+
+  // SCRUBBING CONTROLS ////////////////////////////////////////////////////
+
+  const handleKeyDown = useCallback(
+    ({ key }: KeyboardEvent): void => {
+      if (key === "ArrowLeft" || key === "Left") {
+        timeControls?.handleFrameAdvance(-1);
+      } else if (key === "ArrowRight" || key === "Right") {
+        timeControls?.handleFrameAdvance(1);
+      }
+    },
+    [timeControls]
+  );
+
+  useEffect(() => {
+    window.addEventListener("keydown", handleKeyDown);
+
+    return () => {
+      window.removeEventListener("keydown", handleKeyDown);
+    };
+  }, [handleKeyDown]);
+
+  // Store the current value of the time slider as its own state, and update
+  // the frame using a debounced value to prevent constant updates as it moves.
+  const debouncedFrameInput = useDebounce(frameInput, 250);
+  useEffect(() => {
+    setFrame(debouncedFrameInput);
+  }, [debouncedFrameInput]);
+
+  const handleFindTrack = useCallback(async (): Promise<void> => {
+    // Load track value
+    await findTrack(parseInt(findTrackInput, 10));
+  }, [findTrackInput, findTrack]);
+
+  // RECORDING CONTROLS ////////////////////////////////////////////////////
+
+  // Update the callback for TimeControls and RecordingControls if it changes.
+  // TODO: TimeControls and RecordingControls should be refactored into components
+  // and receive setFrame as props.
+  useMemo(() => {
+    timeControls?.setFrameCallback(setFrame);
+    recordingControls?.setFrameCallback(setFrame);
+  }, [setFrame]);
+
+  // Update the default recording controls prefix.
+  useMemo(() => {
+    if (useDefaultPrefix && datasetName && featureName) {
+      setImagePrefix(`${datasetName}-${featureName}-`);
+    }
+  }, [useDefaultPrefix, datasetName, featureName]);
```

Comment:
Hmm, I think there isn't any particular reason to favor `useMemo` over `useEffect` here. It probably doesn't matter where in the render lifecycle this update happens...

Honestly this is probably a meaningless optimization and it might be fine to run this on every render without caching the results? (timeControls/recordingControls don't do any computation for this beyond assigning a field.)

Reply:
Worth testing, but I think `useEffect` would probably be best here (or no hook at all). I also definitely agree that long term `TimeControls` and `RecordingControls` ought to be available as React components and that `setFrame` ought to be passed down to them as a prop, though I'm unsure how those components ought to look under the hood (per my comment below).

Reply:
Removed the hook here for now.